src/engine/xrinput.js

```diff
@@ -1,136 +1,189 @@
+
         this.inputDebugString = "";
-        this.controllers.forEach((e) =>
+        this.inputSources.forEach((e) =>
         {
-            // console.log(e);
-            e.gamepad.buttons.forEach((f, i) =>
+            e.gamepad.buttons.forEach((button, i) =>
             {
-                if (f.pressed == true)
+                if (button.pressed == true)
                 {
                     this.inputDebugString += (e.handedness + " controller button " + i + "\n");
-                    this.inputDebugString += ("value: " + f.value + "\n");
+                    this.inputDebugString += ("value: " + button.value + "\n");
                 }
             });
 
-            // axes 0 and 1 unused at least on the quest
-            if (e.gamepad.axes[ 2 ] != 0 || e.gamepad.axes[ 3 ] != 0)
+            e.gamepad.axes.forEach((axis, axisIndex) =>
             {
-                this.inputDebugString += e.handedness + " joystick:\n";
-                this.inputDebugString += "x: " + e.gamepad.axes[ 2 ] + "\n";
-                this.inputDebugString += "y: " + e.gamepad.axes[ 3 ] + "\n";
-            }
+                if (axis != 0)
+                {
+                    this.inputDebugString += e.handedness + " joystick:\n";
+
+                    // regardless of axis count, odds will be X, evens will be Y
+                    if (axisIndex % 2 == 0)
+                    {
+                        this.inputDebugString += "x: " + axis + "\n";
+                    }
+                    else // Y (typically 1 or 3)
+                    {
+                        this.inputDebugString += "y: " + axis + "\n";
+                    }
+                }
+            });
         });
-
         return this.inputDebugString == "" ? 0 : console.log(this.inputDebugString);
     }
 }
 
 //xrInput singleton
 const XRInput = new XRInputClass();
 
-// init input on XR session start
+// subscribe to input events on XR session start
 State.eventHandler.addEventListener("xrsessionstarted", (e) =>
 {
-    console.warn("xr session started");
-    State.currentSession = e;
-    State.isXRSession = true;
-
-    //buggy on MC? Should replace arbitrary "2"
-    // const s = Renderer.xr.getSession();
-    // console.log(s.inputSources);
-    for (let i = 0; i < 2; i++)
+    e.addEventListener('selectend', XRInput.onSelectEnd.bind(XRInput));
+    e.addEventListener('selectstart', XRInput.onSelectStart.bind(XRInput));
+    e.addEventListener('select', XRInput.onSelect.bind(XRInput));
+    e.addEventListener('squeezestart', XRInput.onSqueezeStart.bind(XRInput));
+    e.addEventListener('squeezeend', XRInput.onSqueezeEnd.bind(XRInput));
+    e.addEventListener('squeeze', XRInput.onSqueeze.bind(XRInput));
+    e.addEventListener('connected', XRInput.onConnected.bind(XRInput));
+    e.addEventListener('disconnected', XRInput.onDisconnected.bind(XRInput));
+});
+
+State.eventHandler.addEventListener("inputsourceschange", (e) =>
```

Comment:
it can and is (anecdotally, multiple times over a session) and needs to be taken into account. For example see line 118-119 in `ponxr/scene.js`:

```javascript
State.eventHandler.addEventListener("inputsourceschange", e => {
  if (paddles.length != 0) return; //  avoid false positives, i.e headset put down, but paddles already instantiated
```

I'd rather keep these kinds of decisions to the app level and maintain xrinput close to the metal, even if it makes it a little wilder in those ways - but happy to discuss!